Firestore/Protos/nanopb_pretty_printing/generators.py

```diff
@@ -0,0 +1,415 @@
+
+    self._fields = [self._create_field(f, message_desc) for f in
+                    message_desc.fields]
+    # Make sure fields are printed ordered by tag, for consistency with official
+    # proto libraries.
+    self._fields.sort(key=lambda f: f.tag)
+
+  def _create_field(self, field_desc, message_desc):
+    if isinstance(field_desc, nanopb.OneOf):
+      return OneOfPrettyPrintingGenerator(field_desc, message_desc)
+    else:
+      return FieldPrettyPrintingGenerator(field_desc, message_desc)
+
+  def generate_declaration(self):
+    """Generates the declaration of a `ToString()` member function.
+    """
+
+    return '\n' + _indent(1) + 'std::string ToString(int indent = 0) const;\n'
+
+  def generate_definition(self):
+    """Generates the out-of-class definition of a `ToString()` member function.
+    """
+
+    result = _reindent(0, '''\
+        std::string %s::ToString(int indent) const {
+            std::string header = PrintHeader(indent, "%s", this);
+            std::string result;\n\n''') % (
+    self.full_classname, self._short_classname)
+
+    for field in self._fields:
+      result += str(field)
+
+    can_be_empty = all(f.is_primitive or f.is_repeated for f in self._fields)
+    if can_be_empty:
+      result += _reindent(0, '''
+              bool is_root = indent == 0;
+              if (!result.empty() || is_root) {
+                std::string tail = PrintTail(indent);
+                return header + result + tail;
+              } else {
+                return "";
+              }
+          }\n\n''')
+    else:
+      result += _reindent(0, '''
+              std::string tail = PrintTail(indent);
+              return header + result + tail;
+          }\n\n''')
+
+    return result
+
+
+class FieldPrettyPrintingGenerator:
+  """Generates pretty-printing support for a oneof field.
+
+  The generated C++ code will output the field name and value; the output format
+  is the proto text format, suitable for parsing. Unset fields are not printed.
+  Repeated and optional fields are supported.
+
+  Oneofs are not supported; use `OneOfPrettyPrintingGenerator` instead.
+
+  The actual output will be delegated to a C++ function called
+  `PrintPrimitiveField()`, `PrintEnumField()`, or `PrintMessageField()`,
+  according to the field type; the function is expected to be visible at the
+  point of definition.
+  """
+
+  def __init__(self, field_desc, message_desc):
+    """Args:
+      field_desc: nanopb.Field describing this field.
+      message_desc: nanopb.Message describing the message containing this field.
+    """
+
+    self.name = field_desc.name
+    self.tag = field_desc.tag
+
+    self.is_optional = field_desc.rules == 'OPTIONAL' and field_desc.allocation == 'STATIC'
+    self.is_repeated = field_desc.rules == 'REPEATED'
+    self.is_primitive = field_desc.pbtype != 'MESSAGE'
+    self.is_enum = field_desc.pbtype in ['ENUM', 'UENUM']
+
+  def __str__(self):
```

Comment:
I'm not opposed to this but it doesn't seem to fit exactly with the `generate_declaration`/`generate_definition` split above. It would be clearer which phase this is a part of if this were named similarly. This would also have the effect of making it clearer when we're recursing within the generator.

Reply:
Fair enough. Let's leave it as-is.